packages/augur-core/source/contracts/Augur.sol

```diff
@@ -60,10 +60,48 @@ contract Augur is Controlled, IAugur {
     event EscapeHatchChanged(bool isOn);
     event TimestampSet(uint256 newTimestamp);
 
+    mapping(address => bool) private markets;
     mapping(address => bool) private universes;
     mapping(address => bool) private crowdsourcers;
     mapping(address => bool) private shareTokens;
     mapping(address => bool) private auctionTokens;
+    mapping(address => bool) private trustedSender;
+
+    address public uploader;
+    mapping(bytes32 => address) public registry;
+
+    ITime public time;
+
+    constructor() public {
+        uploader = msg.sender;
+    }
+
+    //
+    // Registry
+    //
+
+    function registerContract(bytes32 _key, address _address) public returns (bool) {
+        require(msg.sender == uploader);
+        require(registry[_key] == address(0));
+        registry[_key] = _address;
+        if (_key == "CompleteSets" || _key == "Orders" || _key == "CreateOrder" || _key == "CancelOrder" || _key == "FillOrder" || _key == "Trade" || _key == "ClaimTradingProceeds") {
+            trustedSender[_address] = true;
+        }
+        if (_key == "Time") {
+
+            time = ITime(_address);
+        }
+        return true;
+    }
+
+    function lookup(bytes32 _key) public view returns (address) {
+        return registry[_key];
+    }
+
+    function finishDeployment() public returns (bool) {
```

Comment:
Anyone can call this, potentially breaking an upload process before it is finished?

Reply:
fair point! Will restrict